json/test_simdjson.cpp

```diff
@@ -23,63 +23,52 @@ struct coordinate_t {
   }
 };
 
-string read_file(const string& filename) {
-  ifstream f(filename);
-  if (!f) {
-    return {};
-  }
-  return string(istreambuf_iterator<char>(f),
-                istreambuf_iterator<char>());
-}
-
-coordinate_t calc(const string& text) {
-  dom::parser pj(0);
-  auto allocate_error = pj.allocate(text.size()); // allocate memory for parsing up to p.size() bytes
-  if (allocate_error) {
-    cerr << allocate_error << endl;
-    exit(EXIT_FAILURE);
-  }
-
-  auto [doc, error] = pj.parse(text); // do the parsing, return 0 on success
-  if (error) {
-    cerr << error << endl;
-    exit(EXIT_FAILURE);
-  }
-
-  auto x = 0.0, y = 0.0, z = 0.0;
-  auto len = 0;
-
-  for (auto coord : doc["coordinates"]) {
-    double x_coord = coord["x"];
-    x += x_coord;
 
-    double y_coord = coord["y"];
-    y += y_coord;
-
-    double z_coord = coord["z"];
-    z += z_coord;
+using namespace simdjson;
+using namespace simdjson::builtin;
+
+class on_demand {
+public:
+  bool run(const padded_string &json);
+  coordinate_t my_point{};
+  size_t count{};
+private:
+  ondemand::parser parser{};
+};
 
-    len++;
+bool on_demand::run(const padded_string &json) {
+  count = 0;
+  auto doc = parser.iterate(json);
+  for (ondemand::object point_object : doc["coordinates"]) {
+    my_point.x += double(point_object["x"]);
+    my_point.y += double(point_object["y"]);
+    my_point.z += double(point_object["z"]);
+    count++;
   }
+  return true;
+}
 
-  return coordinate_t(x / len, y / len, z / len);
+coordinate_t calc(const simdjson::padded_string& json) {
+  on_demand reader;
+  reader.run(json);
+  reader.my_point.x /= reader.count;
+  reader.my_point.y /= reader.count;
+  reader.my_point.z /= reader.count;
+  return reader.my_point;
 }
 
 int main() {
-  auto left = calc("{\"coordinates\":[{\"x\":1.1,\"y\":2.2,\"z\":3.3}]}");
+  auto left = calc("{\"coordinates\":[{\"x\":1.1,\"y\":2.2,\"z\":3.3}]}"_padded);
```

Comment:
Could you please clarify why did you use that `_padded` prefix? I guess it would be nice to add the comment to the code regarding that so other code readers won't be confused.

Reply:
I added an explanation. The suffix calls the `simdjson::padded_string` constructor.